Use system's ephemeral port range for port allocation

miminar opened the pull request:
Resolves #10220

This PR causes docker to use system configured ephemeral port range for
allocation of host's free ports to those exported by containers. The range is
read from `/proc/sys/net/ipv4/ip_local_port_range` kernel parameter. If a
number of running containers with exported ports exceeds the size of the range,
the parameter can be used to extend it.

Commit:
Use system's ephemeral port range for port allocation

Read `/proc/sys/net/ipv4/ip_local_port_range` kernel parameter to obtain
ephemeral port range that now sets the boundaries of port allocator
which finds free host ports for those exported by containers.

Signed-off-by: Michal Minar <miminar@redhat.com>

miminar commented:
@rhatdan: Good point, I'll post an update.

jfrazelle commented:
Hey @miminar your PR keeps getting stuck on drone in a race in the tests we fixed today, can you possibly rebase? so sorry.

miminar commented:
@jfrazelle: No problem. The droine's build seems ok now, but I'll rebase nevertheless.

SvenDowideit commented on docs/man/docker-run.1.md:
```diff
@@ -232,8 +232,9 @@ and foreground Docker containers.
 client that can reach the host. When using -P, Docker will bind the exposed
-ports to a random port on the host between 49153 and 65535. To find the
-mapping between the host ports and the exposed ports, use **docker port**.
+port to a random port on the host within an *ephemeral port range* defined by
```
how about `Docker will bind any exposed port to a random ..` as there may be more than one.

miminar replied:
Sounds good.

SvenDowideit commented on docs/sources/articles/networking.md:
```diff
@@ -363,17 +363,19 @@ to provide special options when invoking `docker run`.  These options
-which can be any port at all, not just those in the 49153-65535 block —
-you want mapped to which port in the container.
+First, you can supply `-P` or `--publish-all=true|false` to `docker run` which
+is a blanket operation that identifies every port with an `EXPOSE` line in the
```
or `--expose <port>` commandline flag

miminar replied:
I'm afraid that the `--expose <port>` does not publish container's ports to host.

SvenDowideit replied:
no, but `--expose` is supposed to be equivalent to `EXPOSE` in the Dockerfile

so `--expose 80 --publish-all` should also do what you talk about here.

(assuming i grok the stuff right)

miminar replied:
Oh, I've finally got it :-)

SvenDowideit commented on docs/sources/articles/networking.md:
```diff
@@ -363,17 +363,19 @@ to provide special options when invoking `docker run`.  These options
+First, you can supply `-P` or `--publish-all=true|false` to `docker run` which
+is a blanket operation that identifies every port with an `EXPOSE` line in the
+image's `Dockerfile` and maps it to a host port somewhere within an *ephemeral
+port range*. This tends to be a bit inconvenient, since you then have to run
```
I'd ditch the `inconvenient` phrase, and rewrite that sentence to tell the user to use the `docker port` command to find out what port.

This functionality has important uses that are not about convenience (like running multiple containers that would otherwise conflict)

SvenDowideit commented on docs/sources/articles/networking.md:
```diff
@@ -363,17 +363,19 @@ to provide special options when invoking `docker run`.  These options
+`/proc/sys/net/ipv4/ip_local_port_range` kernel parameter, typically
+ranging from 32768 to 61000.
+
+More convenient is the `-p SPEC` or `--publish=SPEC` option which lets you be
```
and thus change this to something more like `You can explicitly specify the external port ...`

miminar replied:
Will do.

SvenDowideit commented:
Docs LGTM - if you revisit your PR, please consider my comments, but we can also do them later, and you're really just using the existing language.

@fredlf @jamtur01

jamtur01 commented on docs/sources/articles/networking.md:
```diff
@@ -363,17 +363,19 @@ to provide special options when invoking `docker run`.  These options
+image's `Dockerfile` and maps it to a host port somewhere within an *ephemeral
+port range*. This tends to be a bit inconvenient, since you then have to run
+other `docker` sub-commands to learn which external port a given service was
+mapped to. The *ephemeral port range is configured by
```
Formatting incorrect. close the italic.

miminar replied:
Good catch!

jamtur01 commented:
Formatting issues needs fix before Docs are okay.

miminar commented:
@SvenDowideit, @jamtur01: Thanks for a review. Can you take another look?

miminar commented:
@SvenDowideit ping. I'm sorry to bother you. Is it better now?

SvenDowideit commented:
nice :) Docs LGTM

@jfrazelle @erikh is this a proposal? (mmm, yes, I do like the idea that there's a `1-design-review` label :)

fredlf commented on docs/sources/articles/networking.md:
```diff
@@ -363,17 +363,18 @@ to provide special options when invoking `docker run`.  These options
+First, you can supply `-P` or `--publish-all=true|false` to `docker run` which
+is a blanket operation that identifies every port with an `EXPOSE` line in the
+image's `Dockerfile` or `--expose <port>` commandline flag and maps it to a
+host port somewhere within an *ephemeral port range*. `docker port` command
```
Change to "The `docker port` command..." please (as per Style Guide).

fredlf commented on docs/sources/userguide/dockerlinks.md:
```diff
@@ -25,10 +25,10 @@ container that ran a Python Flask application:
-port 5000 in the container was bound to port 49155 on the host.
+When that container was created, the `-P` flag was used to automatically map
+any network port inside it to a random high port within an *ephemeral port
+range* on our Docker host. Next, when `docker ps` was run, you saw that port
```
"...your Docker host..."

fredlf commented:
Some copy edits, otherwise docs LGTM.

miminar commented:
@fredfl: updated. Thanks for review.

fredlf commented:
Thanks for the contribution, @miminar .

gdm85 commented:
Well, I like this more than my original #8412 so hope it makes it! :+1:

LK4D4 commented on daemon/networkdriver/portallocator/portallocator.go:
```diff
@@ -59,6 +69,28 @@ func NewErrPortAlreadyAllocated(ip string, port int) ErrPortAlreadyAllocated {
+		log.Errorf("Failed to read %s kernel parameter: %s", param, err.Error())
+	} else {
+		ports := rePorts.FindAllString(string(line), 2)
+		start, _ = strconv.Atoi(ports[0])
```
If that file will be empty there will be panic. You should check that `len(ports) == 2` and log error if not.

miminar replied:
Ok, better be on a safe side.

LK4D4 commented on daemon/networkdriver/portallocator/portallocator_test.go:
```diff
@@ -5,12 +5,23 @@ import (
+		testPortRangeStart = defaultPortRangeStart
+		testPortRangeEnd   = testPortRangeStart + 500
+	)
+	BeginPortRange = testPortRangeStart
```
I wonder if you can do this in `init` for tests.

miminar replied:
Sure.

jfrazelle commented:
seems reasonable enough to me, wdyt @crosbymichael

icecrime commented on daemon/networkdriver/portallocator/portallocator.go:
```diff
@@ -3,8 +3,23 @@ package portallocator
+
+var (
+	BeginPortRange = 0
+	EndPortRange   = 0
```
It feels a bit counterintuitive that these (mutable) `var`s are exported while the `const` above aren't.

miminar replied:
Yeah, agreed. I just wanted to preserve those exported variables so I didn't have to touch other modules using them. Currently they are used just in test modules though. I'll make it other way around and provide a getter for `(Begin|End)PortRange`.

icecrime commented on daemon/networkdriver/portallocator/portallocator.go:
```diff
@@ -59,6 +69,32 @@ func NewErrPortAlreadyAllocated(ip string, port int) ErrPortAlreadyAllocated {
+	if line, err := ioutil.ReadFile(param); err != nil {
+		log.Errorf("Failed to read %s kernel parameter: %s", param, err.Error())
+	} else {
+		ports := rePorts.FindAllString(string(line), 2)
```
Isn't [`fmt.Fscanf`](https://golang.org/pkg/fmt/#Fscanf) more appropriate than the `regexp` based solution here? [Example](https://play.golang.org/p/eATTtcQpio):

```
if n, err := fmt.Fscanf(strings.NewReader(line), "%d:%d", &start, &end); n != 2 || err != nil {
   [...]
}
```

miminar replied:
Probably yes. I find regexp solution more robust but this is a way simpler. I'll change it this way. Thanks for a tip.

miminar commented:
@icecrime: Could you please take a look once more?

icecrime commented on daemon/networkdriver/portallocator/portallocator_test.go:
```diff
@@ -5,6 +5,11 @@ import (
 
+func init() {
+	beginPortRange = DefaultPortRangeStart
+	endPortRange = DefaultPortRangeStart + 500
```
Agree with @rhatdan, we might as well test the entire range for consistency, but it's not a blocker.

miminar replied:
Alright, changed so that whole default port range is tested.

icecrime commented:
LGTM! @LK4D4 WDYT?

Commit:
Use default port range in unit tests

Signed-off-by: Michal Minar <miminar@redhat.com>

LK4D4 commented on daemon/networkdriver/portallocator/portallocator.go:
```diff
@@ -59,6 +68,29 @@ func NewErrPortAlreadyAllocated(ip string, port int) ErrPortAlreadyAllocated {
+		log.Errorf("Failed to read %s kernel parameter: %s", param, err.Error())
+	} else {
+		var start, end int
+		if n, err := fmt.Fscanf(strings.NewReader(string(line)), "%d\t%d", &start, &end); n != 2 || err != nil {
```
File is already reader. So you should use `os.Open` to open file and pass it to `Fscanf`

miminar replied:
This approach does not work. There's no problem with parsing regular file, but somehow this kernel parameter does not like to be read byte after byte which apparently `fmt.Fscanf` does. Extract from `strace` output:

    open("/proc/sys/net/ipv4/ip_local_port_range", O_RDONLY|O_CLOEXEC) = 3
    futex(0x560290, FUTEX_WAIT, 0, NULL)    = 0
    read(3, "3", 1)                         = 1
    read(3, "", 1)                          = 0
    write(1, "Failed to parse port range from "..., 76Failed to parse port range from /proc/sys/net/ipv4/ip_local_port_range: EOF
    ) = 76
    close(3)                                = 0

But perhaps I'm doing it wrong... Here's my reproducer:

    package main
    
    import (
    	"flag"
    	"fmt"
    	"os"
    )
    
    func main() {
    	var filename = flag.String("file", "/proc/sys/net/ipv4/ip_local_port_range", "File to read numbers from.")
    	flag.Parse()
    
    	file, err := os.Open(*filename)
    	defer file.Close()
    	if err != nil {
    		fmt.Printf("Failed to read file %q: %v\n", filename, err)
    	}
    	var start, end int
    	if n, err := fmt.Fscanf(file, "%d\t%d", &start, &end); n != 2 || err != nil {
    		if err == nil {
    			err = fmt.Errorf("unexpected count of parsed numbers (%d)\n", n)
    		}
    		fmt.Printf("Failed to parse port range from %s: %v\n", *filename, err)
    	} else {
    		fmt.Printf("Parsed range: %d - %d\n", start, end)
    	}
    }

LK4D4 commented on daemon/networkdriver/portallocator/portallocator.go:
```diff
@@ -59,6 +68,29 @@ func NewErrPortAlreadyAllocated(ip string, port int) ErrPortAlreadyAllocated {
+
+	if line, err := ioutil.ReadFile(param); err != nil {
+		log.Errorf("Failed to read %s kernel parameter: %s", param, err.Error())
+	} else {
```
I'm little confused by flow with elses.
I'm prefer if you'd use like:
```
f, err := os.open(param)
if err != nil {
    log.Errorf(...)
    return
}
```
and so on

miminar replied:
OK, no problem.

LK4D4 commented on daemon/networkdriver/portallocator/portallocator.go:
```diff
@@ -59,6 +68,29 @@ func NewErrPortAlreadyAllocated(ip string, port int) ErrPortAlreadyAllocated {
+	}
+}
+
+func GetPortRange() (int, int) {
```
I don't like such helpers, but if you like it at least make it private.

tiborvass replied:
@LK4D4 he calls it from https://github.com/docker/docker/pull/10238/files#diff-b2058005d9cae888c97c7a0a9ce8e797R132

@miminar I would rename it to `PortRange` though.

miminar replied:
Alright, will do.

Commit:
Restructured port range loader

And renamed `GetPortRange` to `PortRange`.

Signed-off-by: Michal Minar <miminar@redhat.com>

miminar commented:
@LK4D4 ping

tiborvass commented:
@LK4D4 @rhatdan are all your concerns addressed?

LK4D4 commented on daemon/networkdriver/portallocator/portallocator.go:
```diff
@@ -59,6 +68,31 @@ func NewErrPortAlreadyAllocated(ip string, port int) ErrPortAlreadyAllocated {
+		return
+	}
+	var start, end int
+	n, err := fmt.Fscanf(strings.NewReader(string(line)), "%d\t%d", &start, &end)
```
I had comment about `*os.File` as reader to `fmt.Fscanf`. You should use `os.Open` to get it.

LK4D4 replied:
Sorry, didn't see your comment.

LK4D4 replied:
You can wrap `os.File` to `bufio.NewReader` when passing to `Fscanf`. Pretty weird behavior :)

miminar replied:
You're right! It works ;-)

miminar replied:
@LK4D4 fixed, thanks!

rhatdan commented:
LGTM

Commit:
Pass buffered file reader to Fscanf instead of string reader

Unless `file` is wrapped with buffered reader, `fmt.Fscanf` will read
just one byte and terminate with `EOF`.

Signed-off-by: Michal Minar <miminar@redhat.com>

miminar commented:
I hope the issues are resolved now. Thanks for all your input.

LK4D4 commented on daemon/networkdriver/portallocator/portallocator.go:
```diff
@@ -59,6 +68,31 @@ func NewErrPortAlreadyAllocated(ip string, port int) ErrPortAlreadyAllocated {
+
+	file, err := os.Open(param)
+	if err != nil {
+		log.Errorf("Failed to read %s kernel parameter: %s", param, err.Error())
```
`err` can be transfered to `%s` directly by `fmt` package, so you don't need `Error()`.

tiborvass replied:
err should be `%v`

LK4D4 commented on daemon/networkdriver/portallocator/portallocator.go:
```diff
@@ -59,6 +68,31 @@ func NewErrPortAlreadyAllocated(ip string, port int) ErrPortAlreadyAllocated {
 }
 
+func init() {
+	const param = "/proc/sys/net/ipv4/ip_local_port_range"
```
pretty bad naming

LK4D4 commented on daemon/networkdriver/portallocator/portallocator.go:
```diff
@@ -59,6 +68,31 @@ func NewErrPortAlreadyAllocated(ip string, port int) ErrPortAlreadyAllocated {
+		if err == nil {
+			err = fmt.Errorf("unexpected count of parsed numbers (%d)", n)
+		}
+		log.Errorf("Failed to parse port range from %s: %v", param, err)
```
Consistency is cool, so use `%s` for err.

tiborvass replied:
disagree, err should be `%v` :)

LK4D4 commented:
@miminar I'm going to merge it. You can fix nits in other PR, pls let me know if you're not interested in it :)

LK4D4 commented:
LGTM

miminar commented:
@LK4D4 thanks! Please check out PR #11402 which fixes your findings.
